src/Microsoft.AspNet.Mvc.Core/ModelBinding/BodyModelBinder.cs

```diff
@@ -28,36 +27,45 @@ public BodyModelBinder()
         protected async override Task<ModelBindingResult> BindModelCoreAsync(
             [NotNull] ModelBindingContext bindingContext)
         {
+            // For compatibility with MVC 5.0 for top level object we want to consider an empty key instead of
+            // the parameter name/a custom name. In all other cases (like when binding body to a property) we
+            // consider the entire ModelName as a prefix.
+            var isTopLevelObject = bindingContext.ModelMetadata.ContainerType == null;
+            var modelBindingKey = isTopLevelObject ? string.Empty : bindingContext.ModelName;
+
             var httpContext = bindingContext.OperationBindingContext.HttpContext;
-            var formatters = bindingContext.OperationBindingContext.InputFormatters;
 
             var formatterContext = new InputFormatterContext(
-                httpContext, 
-                bindingContext.ModelState, 
+                httpContext,
+                bindingContext.ModelState,
                 bindingContext.ModelType);
+            var formatters = bindingContext.OperationBindingContext.InputFormatters;
             var formatter = formatters.FirstOrDefault(f => f.CanRead(formatterContext));
 
             if (formatter == null)
             {
                 var unsupportedContentType = Resources.FormatUnsupportedContentType(
                     bindingContext.OperationBindingContext.HttpContext.Request.ContentType);
-                bindingContext.ModelState.AddModelError(bindingContext.ModelName, unsupportedContentType);
+                bindingContext.ModelState.AddModelError(modelBindingKey, unsupportedContentType);
 
-                // This model binder is the only handler for the Body binding source.
-                // Always tell the model binding system to skip other model binders i.e. return non-null.
-                return new ModelBindingResult(model: null, key: bindingContext.ModelName, isModelSet: false);
+                // This model binder is the only handler for the Body binding source and it cannot run twice. Always
+                // tell the model binding system to skip other model binders and never to fall back i.e. indicate a
+                // fatal error.
+                return new ModelBindingResult(modelBindingKey);
             }
 
             try
             {
+                var previousCount = bindingContext.ModelState.ErrorCount;
                 var model = await formatter.ReadAsync(formatterContext);
```

Comment:
Log an issue to expand `ReadAsync()` or `FormatterContext` to handle
- adding errors at prefixed location in `ModelState`
- lack of a clear indication of an error since at least JSON formatter will never `throw`
- apparent lack of tests

Note `ErrorCount` checks won't handle the corner case where we've already got too many errors in the `ModelState`.